adcc/workflow.py

```diff
@@ -495,3 +509,73 @@ def setup_solver_printing(solmethod_name, matrix, kind, default_print,
         def inner_callback(state, identifier):
             default_print(state, identifier, output)
         return inner_callback
+
+
+def setup_environment(matrix, environment):
+    """
+    Setup environment matrix terms and/or energy corrections.
+    Internal function called from run_adc.
+    """
+    valid_envs = ["ptss", "ptlr", "linear_response"]
+    hf = matrix.reference_state
+    if hf.environment and environment is None:
+        raise InputError(
+            "Environment found in reference state, but no environment"
+            " configuration specified. Please select from the following"
+            f" schemes: {valid_envs}."
+        )
+    elif environment and not hf.environment:
+        raise InputError(
+            "Environment specified, but no environment"
+            " was found in reference state."
+        )
+    elif not hf.environment:
+        environment = False
+
+    if isinstance(environment, bool):
+        environment = {"ptss": True, "ptlr": True} if environment else {}
+    elif isinstance(environment, list):
+        environment = {k: True for k in environment}
+    elif isinstance(environment, str):
+        environment = {environment: True}
+    elif not isinstance(environment, dict):
+        raise TypeError("Invalid type for environment parameter"
+                        f"' {type(environment)}'.")
+
+    if any(env not in valid_envs for env in environment):
+        raise InputError("Invalid key specified for environment."
+                         f" Valid keys are '{valid_envs}'.")
+
+    env_matrix_term = None
+    energy_corrections = []
+
+    forbidden_combinations = [
+        ["ptlr", "linear_response"],
+    ]
+    for fbc in forbidden_combinations:
+        if all(environment.get(k, False) for k in fbc):
+            raise InputError("Combination of environment schemes"
+                             f" '{fbc}' not allowed. Check the"
+                             " adcc documentation for more details.")
+
+    for pt in ["ptss", "ptlr"]:
+        if not environment.get(pt, False):
+            continue
+        hf_corr = hf.excitation_energy_corrections
+        eec_key = f"{hf.environment}_{pt}_correction"
+        if eec_key not in hf_corr:
+            raise ValueError(f"{pt} correction requested, but could not find"
+                             f" the needed function {eec_key} in"
+                             f" reference state from backend {hf.backend}.")
```

Comment:
Show what is available? Or maybe too much?

Reply:
I guess it's a bit too much for this print-out. 😄

Reply:
ok